controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -1236,28 +1225,43 @@ public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenChec
 
     private <ResultT> void sendRequestAsync(final WireCommand request, final ReplyProcessor replyProcessor,
                                             final CompletableFuture<ResultT> resultFuture,
-                                            final ConnectionFactory connectionFactory, final PravegaNodeUri uri) {
-        CompletableFuture<ClientConnection> connectionFuture = connectionFactory.establishConnection(uri, replyProcessor);
+                                            final PravegaNodeUri uri) {
+        // get connection manager for the segment store node from the connectionmanager. 
+        SegmentStoreConnectionManager.SegmentStoreConnectionPool pool = connectionManager.getPool(uri);
+        // take a new connection from the connection manager
+        CompletableFuture<SegmentStoreConnectionManager.ConnectionObject> connectionFuture = pool.getConnection(replyProcessor);
         connectionFuture.whenComplete((connection, e) -> {
-            if (connection == null) {
-                resultFuture.completeExceptionally(new WireCommandFailedException(new ConnectionFailedException(e),
+            if (connection == null || e != null) {
+                ConnectionFailedException cause = e != null ? new ConnectionFailedException(e) : new ConnectionFailedException();
+                resultFuture.completeExceptionally(new WireCommandFailedException(cause,
                         request.getType(),
                         WireCommandFailedException.Reason.ConnectionFailed));
             } else {                
-                connection.sendAsync(request, cfe -> {
-                    if (cfe != null) {
-                        Throwable cause = Exceptions.unwrap(cfe);
-                        if (cause instanceof ConnectionFailedException) {
-                            resultFuture.completeExceptionally(new WireCommandFailedException(cause, request.getType(), WireCommandFailedException.Reason.ConnectionFailed));
-                        } else {
-                            resultFuture.completeExceptionally(new RuntimeException(cause));
-                        }                        
-                    }
-                });                
+                connection.sendAsync(request, resultFuture);
             }
         });
         resultFuture.whenComplete((result, e) -> {
-            connectionFuture.thenAccept(ClientConnection::close);
+            // when processing completes, return the connection back to connection manager asynchronously.
+            // Note: If result future is complete, connectionFuture is definitely complete. 
+            if (resultFuture.isCompletedExceptionally()) {
+                resultFuture.exceptionally(ex -> {
+                    Throwable unwrap = Exceptions.unwrap(ex);
+                    if (unwrap instanceof WireCommandFailedException && 
```

Comment:
This strikes me a somewhat fragile. Could we make this a whitelist instead of a blacklist?

Reply:
i dont quite follow the comment. ?.
The check here is:
if future completed exceptionally with `connectionFailed` or `connectionDropped` then we want to inform the pool about the failure before returning the connection back to it.
else we return the connection as is because the connection is still valid.

Reply:
Yes, I was suggesting we generalize this to include other types of exceptions especially unknown exceptions. If there are exceptions for which we want to consider the connection still valid, then we should have a conditional on those to return the connection.

Reply:
opened issue #3577